<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Query specific variants from specific GWAS — associations • ieugwasr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Query specific variants from specific GWAS — associations"><meta property="og:description" content="Every rsid is searched for against each requested GWAS id. To get a list of
available GWAS ids, or to find their meta data, use gwasinfo.
Can request LD proxies for instances when the requested rsid is not present
in a particular GWAS dataset. This currently only uses an LD reference panel
composed of Europeans in 1000 genomes version 3.
It is also restricted to biallelic single nucleotide polymorphisms (no indels),
with European MAF &amp;gt; 0.01."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ieugwasr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/guide.html">Perform fast queries against a massive database of complete GWAS summary data</a>
    </li>
    <li>
      <a href="../articles/local_ld.html">Running local LD operations</a>
    </li>
    <li>
      <a href="../articles/timings.html">Comparison of rsid vs chr:pos lookup in ElasticSearch</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/MRCIEU/ieugwasr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Query specific variants from specific GWAS</h1>
    <small class="dont-index">Source: <a href="https://github.com/MRCIEU/ieugwasr/blob/HEAD/R/query.R" class="external-link"><code>R/query.R</code></a></small>
    <div class="hidden name"><code>associations.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Every rsid is searched for against each requested GWAS id. To get a list of
available GWAS ids, or to find their meta data, use <code><a href="gwasinfo.html">gwasinfo</a></code>.
Can request LD proxies for instances when the requested rsid is not present
in a particular GWAS dataset. This currently only uses an LD reference panel
composed of Europeans in 1000 genomes version 3.
It is also restricted to biallelic single nucleotide polymorphisms (no indels),
with European MAF &gt; 0.01.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">associations</span><span class="op">(</span></span>
<span>  <span class="va">variants</span>,</span>
<span>  <span class="va">id</span>,</span>
<span>  proxies <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  r2 <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  align_alleles <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  palindromes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  maf_threshold <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  access_token <span class="op">=</span> <span class="fu"><a href="check_access_token.html">check_access_token</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>variants</dt>
<dd><p>Array of variants e.g. <code>c("rs234", "7:105561135-105563135")</code></p></dd>


<dt>id</dt>
<dd><p>Array of GWAS studies to query. See <code><a href="gwasinfo.html">gwasinfo</a></code> for available studies</p></dd>


<dt>proxies</dt>
<dd><p><code>0</code> or (default) <code>1</code> - indicating whether to look for proxies</p></dd>


<dt>r2</dt>
<dd><p>Minimum proxy LD rsq value. Default=<code>0.8</code></p></dd>


<dt>align_alleles</dt>
<dd><p>Try to align tag alleles to target alleles (if <code>proxies = 1</code>).
<code>1</code> = yes (default), <code>0</code> = no</p></dd>


<dt>palindromes</dt>
<dd><p>Allow palindromic SNPs (if <code>proxies = 1</code>). <code>1</code> = yes (default), <code>0</code> = no</p></dd>


<dt>maf_threshold</dt>
<dd><p>MAF threshold to try to infer palindromic SNPs. Default = <code>0.3</code>.</p></dd>


<dt>access_token</dt>
<dd><p>Google OAuth2 access token.
Used to authenticate level of access to data.
By default, checks if already authenticated through <code><a href="get_access_token.html">get_access_token</a></code> and
if not then does not perform authentication</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Dataframe</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gibran Hemani, Ben Elsworth, Tom Palmer, Rita Rasteiro.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

