<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform LD clumping on SNP data — ld_clump • ieugwasr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform LD clumping on SNP data — ld_clump"><meta property="og:description" content="Uses PLINK clumping method, where SNPs in LD within a particular window will be pruned.
The SNP with the lowest p-value is retained."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ieugwasr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/guide.html">Perform fast queries against a massive database of complete GWAS summary data</a>
    </li>
    <li>
      <a href="../articles/local_ld.html">Running local LD operations</a>
    </li>
    <li>
      <a href="../articles/timings.html">Comparison of rsid vs chr:pos lookup in ElasticSearch</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/MRCIEU/ieugwasr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform LD clumping on SNP data</h1>
    <small class="dont-index">Source: <a href="https://github.com/MRCIEU/ieugwasr/blob/HEAD/R/ld_clump.R" class="external-link"><code>R/ld_clump.R</code></a></small>
    <div class="hidden name"><code>ld_clump.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Uses PLINK clumping method, where SNPs in LD within a particular window will be pruned.
The SNP with the lowest p-value is retained.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ld_clump</span><span class="op">(</span></span>
<span>  dat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clump_kb <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  clump_r2 <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  clump_p <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  pop <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>  access_token <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dat</dt>
<dd><p>Dataframe. Must have a variant name column (<code>rsid</code>) and pval column called <code>pval</code>.
If <code>id</code> is present then clumping will be done per unique id.</p></dd>


<dt>clump_kb</dt>
<dd><p>Clumping kb window. Default is very strict, <code>10000</code></p></dd>


<dt>clump_r2</dt>
<dd><p>Clumping r2 threshold. Default is very strict, <code>0.001</code></p></dd>


<dt>clump_p</dt>
<dd><p>Clumping sig level for index variants. Default = <code>1</code> (i.e. no threshold)</p></dd>


<dt>pop</dt>
<dd><p>Super-population to use as reference panel. Default = <code>"EUR"</code>.
Options are <code>"EUR"</code>, <code>"SAS"</code>, <code>"EAS"</code>, <code>"AFR"</code>, <code>"AMR"</code>.
<code>'legacy'</code> also available - which is a previously used verison of the EUR
panel with a slightly different set of markers</p></dd>


<dt>access_token</dt>
<dd><p>Google OAuth2 access token. Used to authenticate level of access to data</p></dd>


<dt>bfile</dt>
<dd><p>If this is provided then will use the API. Default = <code>NULL</code></p></dd>


<dt>plink_bin</dt>
<dd><p>If <code>NULL</code> and <code>bfile</code> is not <code>NULL</code> then will detect
packaged plink binary for specific OS. Otherwise specify path to plink binary.
Default = <code>NULL</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Data frame</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function interacts with the OpenGWAS API, which houses LD reference panels
for the 5 super-populations in the 1000 genomes reference panel.
It includes only bi-allelic SNPs with MAF &gt; 0.01, so it's quite possible that
a variant you want to include in the clumping process will be absent.
If it is absent, it will be automatically excluded from the results.</p>
<p>You can check if your variants are present in the LD reference panel using
<code><a href="ld_reflookup.html">ld_reflookup()</a></code>.</p>
<p>This function does put load on the OpenGWAS servers, which makes life more
difficult for other users. We have implemented a method and made available
the LD reference panels to perform clumping locally, see
<code>ld_clump()</code> and related vignettes for details.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gibran Hemani, Ben Elsworth, Tom Palmer, Rita Rasteiro.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

